src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -327,6 +339,132 @@ func (a *metricsAppender) Finalize() {
 	a.tagEncoder = nil
 }
 
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+) ([]samplesAppender, error) {
+	var (
+		pipelines []metadata.PipelineMetadata
+		appenders []samplesAppender
+	)
+	for _, pipeline := range stagedMetadata.Pipelines {
+		// For pipeline which have tags to augment we generate and send
+		// separate IDs. Other pipelines return the same.
+		pipeline := pipeline
+		if len(pipeline.Tags) == 0 && len(pipeline.GraphitePrefix) == 0 {
+			pipelines = append(pipelines, pipeline)
+			continue
+		}
+
+		tags := augmentTags(a.tags, pipeline.GraphitePrefix, pipeline.Tags, pipeline.AggregationID)
+
+		sm := stagedMetadata
+		sm.Pipelines = []metadata.PipelineMetadata{pipeline}
+
+		appender, err := a.newSamplesAppender(tags, sm)
+		if err != nil {
+			return []samplesAppender{}, err
+		}
+		appenders = append(appenders, appender)
+	}
+
+	if len(pipelines) == 0 {
+		return appenders, nil
+	}
+
+	sm := stagedMetadata
+	sm.Pipelines = pipelines
+	appender, err := a.newSamplesAppender(a.tags, sm)
+	if err != nil {
+		return []samplesAppender{}, err
+	}
+	return append(appenders, appender), nil
+}
+
+func (a *metricsAppender) newSamplesAppender(
+	tags *tags,
+	sm metadata.StagedMetadata,
+) (samplesAppender, error) {
+	a.tagEncoder.Reset()
+	if err := a.tagEncoder.Encode(tags); err != nil {
+		return samplesAppender{}, err
+	}
+	data, ok := a.tagEncoder.Data()
+	if !ok {
+		return samplesAppender{}, fmt.Errorf("unable to encode tags: names=%v, values=%v", tags.names, tags.values)
+	}
+	// NB (@shreyas): if this is the last use of the tagEncoder we do not need
+	// this copy, but its hard to know and enforce that.
+	id := make([]byte, data.Len())
```

Comment:
This seems like will add a lot more allocs in the general case, why do we know this is not the last usage?

Reply:
I guess I can pass in a flag for the last usage.. So that should be the general case.